File: Asssingment 2/circular-q.c
//Write a C program to implement circular queue.

#include<stdio.h>
#include<stdlib.h>


struct node
{
    int data;
    struct node *next;
};


struct node *top = NULL;
struct node *bottom = NULL;


void push(int data)
{
    struct node *newNode = (struct node *)malloc(sizeof(struct node));
    newNode->data = data;
    newNode->next = NULL;
    if (top == NULL)
    {
        top = newNode;
        bottom = newNode;
    }
    else
    {
        bottom->next = newNode;
        bottom = newNode;
    }
}

void pop()
{
    if (top == NULL)
    {
        printf("Stack is empty\n");
        return;
    }
    struct node *temp = top;
    top = top->next;
    free(temp);
}

void display()
{
    if (top == NULL)
    {
        printf("Stack is empty\n");
        return;
    }
    struct node *temp = top;
    while (temp != NULL)
    {
        printf("%d ", temp->data);
        temp = temp->next;
    }
    printf("\n");
}

void reverse()
{
    if (top == NULL)
    {
        printf("Stack is empty\n");
        return;
    }
    struct node *temp = top;
    struct node *prev = NULL;
    struct node *next = NULL;
    while (temp != NULL)
    {
        next = temp->next;
        temp->next = prev;
        prev = temp;
        temp = next;
    }
    top = prev;
}

int main()
{
    printf("Enter the number of elements\n");
    int n;
    scanf("%d", &n);
    printf("Enter the elements\n");
    for (int i = 0; i < n; i++)
    {
        int data;
        scanf("%d", &data);
        push(data);
    }
    printf("The elements in the circular queue are\n");
    display();
    printf("The elements in the circular queue in reverse order are\n");
    reverse();
    display();
    return 0;
}

File: Asssingment 2/Python/circular-q.py
# write a python program to implement a circular queue

class CircularQueue:
    
    #Constructor
    def __init__(self):
        self.queue = list()
        self.front = 0
        self.rear = 0
        self.maxSize = 8

    #Adding elements to the queue
    def enqueue(self,data):
        if self.size() == self.maxSize-1:
            return ("Queue Full!")
        self.queue.append(data)
        self.rear = (self.rear + 1) % self.maxSize
        return True

    #Removing elements from the queue
    def dequeue(self):
        if self.size()==0:
            return ("Queue Empty!") 
        data = self.queue[self.front]
        self.front = (self.front + 1) % self.maxSize
        return data

    #Calculating the size of the queue
    def size(self):
        if self.rear>=self.front:
            return (self.rear-self.front)
        return (self.maxSize - (self.front-self.rear))

q = CircularQueue()
print(q.enqueue(1))
print(q.enqueue(2))
print(q.enqueue(3)) 
print(q.enqueue(4))
print(q.enqueue(5))
print(q.dequeue())
print(q.dequeue())
print(q.dequeue())
print(q.dequeue())